### PR TITLE
backend/feat/subscription_model

### PR DESCRIPTION
**Pull request by Anuragini-2506**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Payment.hs [hunk 1]
```diff
@@ -142,24 +105,82 @@ juspayWebhookHandler merchantShortId authData value = do
...
+          case osr of
+            Payment.OrderStatusResp {..} -> do
+              unless (transactionStatus /= Payment.CHARGED) $ do
+                order <- runInReplica $ QOrder.findByShortId (ShortId orderShortId) >>= fromMaybeM (PaymentOrderNotFound orderShortId)
+                processPayment merchantId (cast order.personId) order.id
+            Payment.MandateOrderStatusResp {..} -> do
+              unless (transactionStatus /= Payment.CHARGED) $ do
+                order <- runInReplica $ QOrder.findByShortId (ShortId orderShortId) >>= fromMaybeM (PaymentOrderNotFound orderShortId)
+                processPayment merchantId (cast order.personId) order.id
+                processMandate (cast order.personId) DM.INACTIVE mandateStartDate mandateEndDate mandateId mandateMaxAmount
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Payment.hs:121]: Why `INACTIVE`?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Payment.hs [hunk 2]
```diff
@@ -142,24 +105,82 @@ juspayWebhookHandler merchantShortId authData value = do
...
+      QDFS.clearPaymentStatus driverId driverInfo.active
+
+processMandate :: Id DP.Person -> DM.MandateStatus -> UTCTime -> UTCTime -> Text -> HighPrecMoney -> Flow ()
+processMandate driverId mandateStatus startDate endDate mandateId maxAmount = do
+  Redis.whenWithLockRedis (paymentProcessingLockKey mandateId) 60 $ do
+    mbExistingMandate <- runInReplica $ QM.findById (Id mandateId)
+    case mbExistingMandate of
+      Just mandate -> Esq.runNoTransaction $ QM.updateStatus mandate.id mandateStatus
+      Nothing -> Esq.runTransaction $ QM.create =<< mkMandate
+  Redis.whenWithLockRedis (paymentProcessingLockKey driverId.getId) 60 $ do
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Payment.hs:150]: This lock is on line 144 also

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Payment.hs [hunk 3]
```diff
@@ -142,24 +105,82 @@ juspayWebhookHandler merchantShortId authData value = do
...
+      QM.updateStatus (Id mandateId) DM.ACTIVE
+      QDP.updatePaymentModeByDriverId (cast driverPlan.driverId) DP.AUTOPAY
+    CDI.updateSubscription True (cast driverPlan.driverId)
+    CDI.updateAutoPayStatus (castAutoPayStatus mandateStatus) (cast driverPlan.driverId)
+  when (mandateStatus `elem` [Payment.REVOKED, Payment.FAILURE, Payment.EXPIRED, Payment.PAUSED]) $ do
+    driverPlan <- runInReplica $ QDP.findByMandateId (Id mandateId) >>= fromMaybeM (NoCurrentPlanForDriver mandateId)
+    Esq.runTransaction $ do
+      QM.updateStatus (Id mandateId) DM.INACTIVE
+      QDP.updatePaymentModeByDriverId (cast driverPlan.driverId) DP.MANUAL
+    CDI.updateAutoPayStatus (castAutoPayStatus mandateStatus) (cast driverPlan.driverId)
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Payment.hs:178]: This code will make subscription true even if mandate is not active.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+      Just DI.SUSPENDED -> MANUAL
+      Just DI.PAUSED_PSP -> MANUAL
+      Just DI.CANCELLED_PSP -> MANUAL
+      _ -> MANUAL
+
+-- This API is to create a mandate order if the driver has not subscribed to Mandate even once or has Cancelled Mandate from PSP App.
+planSubscribe :: Id Plan -> (Id SP.Person, Id DM.Merchant) -> Flow PlanSubscribeRes
+planSubscribe planId (driverId, merchantId) = do
+  driverInfo <- CDI.findById (cast driverId) >>= fromMaybeM (PersonNotFound driverId.getId)
+  unless (isNothing driverInfo.autoPayStatus || driverInfo.autoPayStatus == Just DI.CANCELLED_PSP) $ throwError InvalidAutoPayStatus
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs:126]: How will we know if driver cancels through PSP? Shouldn't we fetch the mandate status once before this line?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+------------------------------------------ Helper Functions ---------------------------------------------
+---------------------------------------------------------------------------------------------------------
+
+validateActiveMandateExists :: Id SP.Person -> DriverPlan -> Flow DM.Mandate
+validateActiveMandateExists driverId driverPlan = do
+  case driverPlan.mandateId of
+    Nothing -> throwError $ ActiveMandateDoNotExist driverId.getId
+    Just mandateId -> do
+      mandate <- Esq.runInReplica $ QM.findById mandateId >>= fromMaybeM (MandateNotFound mandateId.getId)
+      unless (mandate.status == DM.INACTIVE) $ throwError (ActiveMandateDoNotExist driverId.getId)
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs:193]: should be `ACTIVE`?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+      return mandate
+
+validateInActiveMandateExists :: Id SP.Person -> DriverPlan -> Flow DM.Mandate
+validateInActiveMandateExists driverId driverPlan = do
+  case driverPlan.mandateId of
+    Nothing -> throwError $ ActiveMandateDoNotExist driverId.getId
+    Just mandateId -> do
+      mandate <- Esq.runInReplica $ QM.findById mandateId >>= fromMaybeM (MandateNotFound mandateId.getId)
+      unless (mandate.status == DM.ACTIVE) $ throwError (ActiveMandateDoNotExist driverId.getId)
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs:202]: should be `INACTIVE`? Error should also change

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+          frequency = Payment.ASPRESENTED
+        }
+    mkDriverFee = do
+      id <- generateGUID
+      now <- getCurrentTime
+      return $
+        DF.DriverFee
+          { id = id,
+            merchantId = merchantId,
+            payBy = now,
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs:236]: The rule is to make driver unsubscribed if they don't pay after `payBy`. Are we not following that here?

Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plan.hs:236]: No @0utkarsh For now while rolling out this feature we will not generate invoices for drivers. That will start after september 1st.

**Pull request closed without merging**